clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetsRequestManager.java

```diff
@@ -218,18 +490,17 @@ public CompletableFuture<Void> resetPositionsIfNeeded() {
      *
      * <p/>
      *
-     * When a response is received, positions are validated and, if a log truncation is
-     * detected, a {@link LogTruncationException} will be saved in memory, to be thrown on the
+     * When a response is received, positions are validated and, if a log truncation is detected, a
+     * {@link LogTruncationException} will be saved in memory in cachedUpdatePositionsException, to be thrown on the
      * next call to this function.
      */
-    public CompletableFuture<Void> validatePositionsIfNeeded() {
-        Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate =
-                offsetFetcherUtils.getPartitionsToValidate();
+    void validatePositionsIfNeeded() {
+        Map<TopicPartition, SubscriptionState.FetchPosition> partitionsToValidate = offsetFetcherUtils.getPartitionsToValidate();
         if (partitionsToValidate.isEmpty()) {
```

Comment:
You're totally right that here we could have empty partitions to validate (only because we may have some but there's already a `OffsetsForLeaderEpochRequest` in-flight for them). But my expectation is that even in that case, we make sure that all partitions end up being validated because of how we handle the OffsetsForLeaderEpochRequest response :
- if it succeeds, partitions get validated
- if it fails, the next allowed retry is updated for those partitions, so they are validated on the next poll/updateFetchPositions after the backoff expires (`partitionsToValidate` won't be empty on that next run)
https://github.com/apache/kafka/blob/02e3f7cc284d062adad5324d9493f39559346ae7/clients/src/main/java/org/apache/kafka/clients/consumer/internals/OffsetFetcherUtils.java#L396

Makes sense?